### ISSUE TITLE
Two workflow related questions

### ISSUE THREAD

**Issue by min-xu-ai** [facebookresearch/fairscale#15]
## ❓ Questions and Help

These are not with high priority. When I was working on PR #14, I have these two issues around workflow remain unresolved. File them here in case we want to resolve.

1. requirements-dev.txt and requirements-test.txt are a little bit confusing. Unclear which one a developer should be using. I suggest we either merge them or document the use cases.
1. when merging a PR one have three choices: merge, squash, and rebase. It is not clear which one we should be using. Merge doesn't change any commit history but one can't modify their commit msg. Squash and rebase could make history more linear and with fewer small commits. But it changes commit history and may render older GitHub review comments hard to find. I don't know if we should recommend one method from the above to make it consistent.

**Comment by min-xu-ai** [comment]
with the latest c++ code addition, I had trouble setting up my env to compile it correctly. I got error like:

```
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
[1/1] /public/apps/cuda/10.1/bin/nvcc -I/scratch/m1n/miniconda3/envs/env1/lib/python3.8/site-packages/torch/include -I/scratch/m1n/miniconda3/envs/env1/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -I/scratch/m1n/miniconda3/envs/env1/l
ib/python3.8/site-packages/torch/include/TH -I/scratch/m1n/miniconda3/envs/env1/lib/python3.8/site-packages/torch/include/THC -I/public/apps/cuda/10.1/include -I/scratch/m1n/miniconda3/envs/env1/include/python3.8 -c -c /private/home/m1n/git/fairscale/fai
rscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu -o /private/home/m1n/git/fairscale/build/temp.linux-x86_64-3.8/fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.o -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPE
RATORS__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -O3 --use_fast_math -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=fused_adam_cuda -D_GLIBCXX_USE_CXX11_ABI=0 -gencode=arch=compute_60,code=sm_60 -std=c++14
FAILED: /private/home/m1n/git/fairscale/build/temp.linux-x86_64-3.8/fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.o
/public/apps/cuda/10.1/bin/nvcc -I/scratch/m1n/miniconda3/envs/env1/lib/python3.8/site-packages/torch/include -I/scratch/m1n/miniconda3/envs/env1/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -I/scratch/m1n/miniconda3/envs/env1/lib/pyt
hon3.8/site-packages/torch/include/TH -I/scratch/m1n/miniconda3/envs/env1/lib/python3.8/site-packages/torch/include/THC -I/public/apps/cuda/10.1/include -I/scratch/m1n/miniconda3/envs/env1/include/python3.8 -c -c /private/home/m1n/git/fairscale/fairscale
/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu -o /private/home/m1n/git/fairscale/build/temp.linux-x86_64-3.8/fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.o -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS
__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -O3 --use_fast_math -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=fused_adam_cuda -D_GLIBCXX_USE_CXX11_ABI=0 -gencode=arch=compute_60,code=sm_60 -std=c++14
/usr/include/c++/7/bits/basic_string.tcc: In instantiation of ‘static std::basic_string<_CharT, _Traits, _Alloc>::_Rep* std::basic_string<_CharT, _Traits, _Alloc>::_Rep::_S_create(std::basic_string<_CharT, _Traits, _Alloc>::size_type, std::basic_string<_
CharT, _Traits, _Alloc>::size_type, const _Alloc&) [with _CharT = char16_t; _Traits = std::char_traits<char16_t>; _Alloc = std::allocator<char16_t>; std::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]’:
/usr/include/c++/7/bits/basic_string.tcc:578:28:   required from ‘static _CharT* std::basic_string<_CharT, _Traits, _Alloc>::_S_construct(_InIterator, _InIterator, const _Alloc&, std::forward_iterator_tag) [with _FwdIterator = const char16_t*; _CharT = c
har16_t; _Traits = std::char_traits<char16_t>; _Alloc = std::allocator<char16_t>]’
/usr/include/c++/7/bits/basic_string.h:5042:20:   required from ‘static _CharT* std::basic_string<_CharT, _Traits, _Alloc>::_S_construct_aux(_InIterator, _InIterator, const _Alloc&, std::__false_type) [with _InIterator = const char16_t*; _CharT = char16_
t; _Traits = std::char_traits<char16_t>; _Alloc = std::allocator<char16_t>]’
/usr/include/c++/7/bits/basic_string.h:5063:24:   required from ‘static _CharT* std::basic_string<_CharT, _Traits, _Alloc>::_S_construct(_InIterator, _InIterator, const _Alloc&) [with _InIterator = const char16_t*; _CharT = char16_t; _Traits = std::char_
traits<char16_t>; _Alloc = std::allocator<char16_t>]’
/usr/include/c++/7/bits/basic_string.tcc:656:134:   required from ‘std::basic_string<_CharT, _Traits, _Alloc>::basic_string(const _CharT*, std::basic_string<_CharT, _Traits, _Alloc>::size_type, const _Alloc&) [with _CharT = char16_t; _Traits = std::char_
traits<char16_t>; _Alloc = std::allocator<char16_t>; std::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]’
/usr/include/c++/7/bits/basic_string.h:6688:95:   required from here
/usr/include/c++/7/bits/basic_string.tcc:1067:16: error: cannot call member function ‘void std::basic_string<_CharT, _Traits, _Alloc>::_Rep::_M_set_sharable() [with _CharT = char16_t; _Traits = std::char_traits<char16_t>; _Alloc = std::allocator<char16_t
>]’ without object
       __p->_M_set_sharable();
       ~~~~~~~~~^~
/usr/include/c++/7/bits/basic_string.tcc: In instantiation of ‘static std::basic_string<_CharT, _Traits, _Alloc>::_Rep* std::basic_string<_CharT, _Traits, _Alloc>::_Rep::_S_create(std::basic_string<_CharT, _Traits, _Alloc>::size_type, std::basic_string<_CharT, _Traits, _Alloc>::size_type, const _Alloc&) [with _CharT = char32_t; _Traits = std::char_traits<char32_t>; _Alloc = std::allocator<char32_t>; std::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]’:
/usr/include/c++/7/bits/basic_string.tcc:578:28:   required from ‘static _CharT* std::basic_string<_CharT, _Traits, _Alloc>::_S_construct(_InIterator, _InIterator, const _Alloc&, std::forward_iterator_tag) [with _FwdIterator = const char32_t*; _CharT = char32_t; _Traits = std::char_traits<char32_t>; _Alloc = std::allocator<char32_t>]’
/usr/include/c++/7/bits/basic_string.h:5042:20:   required from ‘static _CharT* std::basic_string<_CharT, _Traits, _Alloc>::_S_construct_aux(_InIterator, _InIterator, const _Alloc&, std::__false_type) [with _InIterator = const char32_t*; _CharT = char32_t; _Traits = std::char_traits<char32_t>; _Alloc = std::allocator<char32_t>]’
/usr/include/c++/7/bits/basic_string.h:5063:24:   required from ‘static _CharT* std::basic_string<_CharT, _Traits, _Alloc>::_S_construct(_InIterator, _InIterator, const _Alloc&) [with _InIterator = const char32_t*; _CharT = char32_t; _Traits = std::char_traits<char32_t>; _Alloc = std::allocator<char32_t>]’
/usr/include/c++/7/bits/basic_string.tcc:656:134:   required from ‘std::basic_string<_CharT, _Traits, _Alloc>::basic_string(const _CharT*, std::basic_string<_CharT, _Traits, _Alloc>::size_type, const _Alloc&) [with _CharT = char32_t; _Traits = std::char_traits<char32_t>; _Alloc = std::allocator<char32_t>; std::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]’
/usr/include/c++/7/bits/basic_string.h:6693:95:   required from here
/usr/include/c++/7/bits/basic_string.tcc:1067:16: error: cannot call member function ‘void std::basic_string<_CharT, _Traits, _Alloc>::_Rep::_M_set_sharable() [with _CharT = char32_t; _Traits = std::char_traits<char32_t>; _Alloc = std::allocator<char32_t>]’ without object
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "/scratch/m1n/miniconda3/envs/env1/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1418, in _run_ninja_build
    subprocess.run(
  File "/scratch/m1n/miniconda3/envs/env1/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ninja', '-v']' returned non-zero exit status 1.
```

It seems that we need better instruction on how to setup things up, esp for:

1. making sure pytorch with cuda can be installed from requirements.txt
2. making sure cpp dev env is correct

**Comment by min-xu-ai** [comment]
> 2. when merging a PR one have three choices: merge, squash, and rebase. It is not clear which one we should be using. Merge doesn't change any commit history but one can't modify their commit msg. Squash and rebase could make history more linear and with fewer small commits. But it changes commit history and may render older GitHub review comments hard to find. I don't know if we should recommend one method from the above to make it consistent.

This is resolved. We have decided on squash & rebase as the way going forward.

**Comment by min-xu-ai** [comment]
@msbaines is this worth keeping open for a little while?

**Comment by msbaines** [comment]
> 1. making sure pytorch with cuda can be installed from requirements.txt

This should be addressed by PR #21
